NEW: Correctly expose Media keys through keyboard

todi1856 opened the pull request:
### Description

See https://github.cds.internal.unity3d.com/unity/unity/pull/74515 for details


### Testing status & QA

Added coverage in unity/unity repo, also manually tested media keys on FireTV using Media.zip from https://jira.unity3d.com/browse/UUM-113004

### Overall Product Risks


- Complexity:  Low
- Halo Effect:  Low

### Comments to reviewers

_Please describe any additional information such as what to focus on, or historical info for the reviewers._

### Checklist

Before review:

- [X] Changelog entry added.
    - Explains the change in `Changed`, `Fixed`, `Added` sections.
    - For API change contains an example snippet and/or migration example.
    - JIRA ticket linked, example ([case %<ID>%](https://issuetracker.unity3d.com/product/unity/issues/guid/<ID>)). If it is a private issue, just add the case ID without a link.
    - Jira port for the next release set as "Resolved".
- [X] Tests added/changed, if applicable.
    - Functional tests `Area_CanDoX`, `Area_CanDoX_EvenIfYIsTheCase`, `Area_WhenIDoX_AndYHappens_ThisIsTheResult`.
    - Performance tests.
    - Integration tests.
- [X] Docs for new/changed API's.
    - Xmldoc cross references are set correctly.
    - Added explanation how the API works.
    - Usage code examples added.
    - The manual is updated, if needed.

todi1856 committed:
expose media keys

todi1856 committed:
docs

todi1856 committed:
changelog

todi1856 committed:
Add missing

todi1856 committed:
bump

Pauliusd01 committed:
Merge branch 'develop' into expose-media-keys

jfreire-unity commented on Packages/com.unity.inputsystem/InputSystem/Devices/Precompiled/FastKeyboard.cs:
```diff
@@ -2075,7 +2087,7 @@ private UnityEngine.InputSystem.Controls.KeyControl Initialize_ctrlKeyboardright
                 })
                 .WithMinAndMax(0, 1)
                 .Finish();
-            ctrlKeyboardrightAlt.keyCode = UnityEngine.InputSystem.Key.RightAlt;
```
Any reason why this and `LeftWindows` changed? Is it because you're on macOS?

todi1856 replied:
I generated this file on Windows, honestly not sure.

todi1856 replied:
Will try on pure develop branch, maybe somebody forgot to update this file ?

todi1856 replied:
Hmph, it's indeed somehow related to my changes, after locally reverting Media Keys, this place reverts itself

todi1856 replied:
@jfreire-unity this feels like a bug in input system.

This is where this place is generated 
https://github.com/Unity-Technologies/InputSystem/blob/e83bf4fd8387210412b9ee5f0861418d32504f18/Packages/com.unity.inputsystem/InputSystem/Editor/InputLayoutCodeGenerator.cs#L390

https://github.com/Unity-Technologies/InputSystem/blob/e83bf4fd8387210412b9ee5f0861418d32504f18/Packages/com.unity.inputsystem/InputSystem/Devices/Keyboard.cs#L578

AltGr and RightAlt are the same values... that's why when string interpolation kicks in, it picks one or another. At least that's how I understand it.

todi1856 replied:
<img width="876" height="267" alt="stringinterpolation" src="https://github.com/user-attachments/assets/f23b984c-23e4-4f48-8d04-14a1d3ca1b1b" />

look here, tt value shows the generated string with AltGr, even though control.keyCode was RightAlt

jfreire-unity replied:
Interesting, I wonder why this is different from your changes. I'll have a look but I'll approve since tests are passing and it's not yet clear to me the impact of this change.
There's still some CI failures due to API validation that need to be fixed

jfreire-unity commented on Packages/com.unity.inputsystem/InputSystem/Devices/Keyboard.cs:
```diff
@@ -969,6 +972,21 @@ public enum Key
+        /// <summary>
+        /// The <see cref="Keyboard.mediaForward"/>.
+        /// </summary>
+        MediaForward,
```
I think we need to change this line from `F24` to `MediaForward` https://github.com/Unity-Technologies/InputSystem/blob/e83bf4fd8387210412b9ee5f0861418d32504f18/Packages/com.unity.inputsystem/InputSystem/Devices/Keyboard.cs#L1051

todi1856 replied:
Fixed https://github.com/Unity-Technologies/InputSystem/pull/2212/commits/a9e26e55b6bed104a36988f8d2c0b70be9b103fc

The layout also changed https://github.com/Unity-Technologies/InputSystem/pull/2212/commits/788544bb20c20ad685db3a3888a8b64a4fc37e5a

jfreire-unity reviewed (commented):
👍 
I compared with latest https://github.com/Unity-Technologies/InputSystem/pull/2075 PR that also added more keys and only found one thing missing which I commented.

todi1856 committed:
review fix

todi1856 committed:
upadate layout

todi1856 committed:
Merge branch 'expose-media-keys' of github.com:Unity-Technologies/InputSystem into expose-media-keys

jfreire-unity reviewed (approved):
Approving but CI errors need to be addressed

todi1856 committed:
Merge branch 'develop' into expose-media-keys

todi1856 committed:
Try bumping to 1.15.0

todi1856 committed:
Merge branch 'expose-media-keys' of github.com:Unity-Technologies/InputSystem into expose-media-keys

jfreire-unity commented:
> Approving but CI errors need to be addressed

Ok it's probably something to be done once we release 1.14.2, we need to bump to 1.15 since we're adding new APIs.

todi1856 commented:
> > Approving but CI errors need to be addressed
> 
> Ok it's probably something to be done once we release 1.14.2, we need to bump to 1.15 since we're adding new APIs.

Alright, I'll wait for 1.15 bump. Thx !

todi1856 committed:
move items in changelog to better reflect things

todi1856 committed:
Update ExtendedSubmitCancelEventData.cs.meta

todi1856 committed:
Update INavigationEventData.cs.meta

todi1856 committed:
revert

Pauliusd01 committed:
Merge branch 'develop' into expose-media-keys

todi1856 committed:
Revert accidental

dri-richard commented:
I submitted https://issuetracker.unity3d.com/issues/android-tv-some-of-the-remote-controls-not-detected-with-inputsystem-dot-devices which I think this PR is designed to address, so I thought I'd try this branch out.

Do you have any guidance on how to use it? I added bindings for:
MediaRewind[ Keyboard]
MediaForward [Keyboard]
MediaPlayPause [Keyboard]
ContextMenu [Keyboard]

and added logging of any action being performed, and of these 4 only ContextMenu was triggered when using a Fire TV remote.

todi1856 commented:
> I submitted https://issuetracker.unity3d.com/issues/android-tv-some-of-the-remote-controls-not-detected-with-inputsystem-dot-devices which I think this PR is designed to address, so I thought I'd try this branch out.
> 
> Do you have any guidance on how to use it? I added bindings for: MediaRewind[ Keyboard] MediaForward [Keyboard] MediaPlayPause [Keyboard] ContextMenu [Keyboard]
> 
> and added logging of any action being performed, and of these 4 only ContextMenu was triggered when using a Fire TV remote.

@dri-richard this PR only covers the frontend, the backend changes (where Unity forwards Android input data to input system package) are in internal PR, not accessible publicly. When the internal PR for backend will land, the status in https://issuetracker.unity3d.com/issues/android-tv-some-of-the-remote-controls-not-detected-with-inputsystem-dot-devices will change. Internal PR has to separately land in affected Unity versions - 6000.3, 6000.2, etc

todi1856 commented:
@jfreire-unity @Pauliusd01 iOS tests are failing on develop branch too - https://unity-ci.cds.internal.unity3d.com/project/130/branch/develop/jobDefinition/.yamato%2Fupm-ci.yml%23run_ios_trunk_functional/recent-jobs, I as

jfreire-unity committed:
Merge branch 'develop' into expose-media-keys

jfreire-unity commented:
@todi1856 yeah we are aware (also, trunk tests are actually not required for PRs to land in our repo).
I now pulled the latest changes from `develop` to this PR to run CI again. if that's the only test that fails, we can merge this PR.

Pauliusd01 commented:
Sorry was busy with release, taking a look today

Pauliusd01 commented:
Since I do need the editor change to actually use these, do we want any kind of documentation or comments to let users know which editors support it? Since IntelliSense will suggest them, users might try it and it will silently fail

Pauliusd01 commented:
Other than my comment above, everything seems fine to me. Checked functionality for the added keys and that F24 key still works as well as there was a related change to it. I can approve this, just let me know

todi1856 commented:
> Since I do need the editor change to actually use these, do we want any kind of documentation or comments to let users know which editors support it? Since IntelliSense will suggest them, users might try it and it will silently fail

I don't have this information yet, as per https://jira.unity3d.com/browse/UUM-113004 it will land in 2022.3, 6000.0, 6000.1, 6000.2, 6000.3, but I don't know the exact versions yet.

Pauliusd01 commented:
> > Since I do need the editor change to actually use these, do we want any kind of documentation or comments to let users know which editors support it? Since IntelliSense will suggest them, users might try it and it will silently fail
> 
> I don't have this information yet, as per https://jira.unity3d.com/browse/UUM-113004 it will land in 2022.3, 6000.0, 6000.1, 6000.2, 6000.3, but I don't know the exact versions yet.

That's good, matches the package release so no need for anything extra then

todi1856 commented:
@Pauliusd01 what now, will somebody from Input team needs to click Merge button ?

P.S I don't have authorization to click Merge.
